core/3d/CCBundle3DData.h

```diff
@@ -39,8 +39,140 @@
+    { 
+        _buffer.clear(); 
+        if (format != backend::IndexFormat::UNSPEC)
+            _stride = format == backend::IndexFormat::U_SHORT ? 2 : 4;
+    }
+
+    /** Pushes back a value if type unsigned short (uint16_t). */
+    void push_back(uint16_t val)
+    {
+        assert(_stride == 2);
+        _buffer.append_n((uint8_t*)&val, sizeof(val));
+    }
+
+    /** Pushes back a value if type unsigned int (uint32_t). */
+    void push_back(uint32_t val, std::true_type /*U_INT*/)
+    {
+        assert(_stride == 4);
+        _buffer.append_n((uint8_t*)&val, sizeof(val));
+    }
+
+    /** Inserts a list containing unsigned short (uint16_t) data. */
+    void insert(uint8_t* where, std::initializer_list<uint16_t> ilist)
+    {
+        assert(_stride == 2);
+        _buffer.insert(where, (uint8_t*)ilist.begin(), (uint8_t*)ilist.end());
+    }
+
+    /** Inserts a list containing unsigned int (uint32_t) data. */
+    void insert(uint8_t* where, std::initializer_list<uint32_t> ilist, std::true_type)
+    {
+        assert(_stride == 4);
+        _buffer.insert(where, (uint8_t*)ilist.begin(), (uint8_t*)ilist.end());
+    }
+
+    template <typename _Ty>
+    _Ty& at(size_t idx)
+    {
+        assert((sizeof(_Ty) == sizeof(uint16_t) && _stride == 2) ||
+               (sizeof(_Ty) == sizeof(uint32_t) && _stride == 4));
+        if (idx < this->size())
+            return (_Ty&)_buffer[idx * sizeof(_Ty)];
+        throw std::out_of_range("IndexArray: out of range!");
+    }
+
+    uint8_t* begin() noexcept { return _buffer.begin(); }
+    uint8_t* end() noexcept { return _buffer.end(); }
+    const uint8_t* begin() const noexcept { return _buffer.begin(); }
+    const uint8_t* end() const noexcept { return _buffer.end(); }
+
+    uint8_t* data() noexcept { return _buffer.data(); }
+    const uint8_t* data() const noexcept { return _buffer.data(); }
+
+    /** returns the count of indices in the container. */
+    size_t size() const { return _buffer.size() / _stride; }
+    /** returns the size of the container in bytes. */
+    size_t bsize() const { return _buffer.size(); }
+
+    /** resizes the count of indices in the container. */
+    void resize(size_t size) { _buffer.resize(size * _stride); }
+    /** resizes the container in bytes. */
+    void bresize(size_t size) { _buffer.resize(size); }
+
+    /** returns true if the container is empty. Otherwise, false. */
+    bool empty() const { return _buffer.empty(); }
+
+    /** returns the format of the index array. */
+    backend::IndexFormat format() const
+    {
+        return _stride == 2 ? backend::IndexFormat::U_SHORT : backend::IndexFormat::U_INT;
+    }
+
+    /** sets the format of the index array. */
+    void format(backend::IndexFormat format)
+    {
+        _stride = format == backend::IndexFormat::U_SHORT ? 2 : 4;
+    }
+
+    template <typename _Fty>
+    void for_each(_Fty cb) const
+    {
+        assert(_stride != 0);
```

Comment:
```suggestion
        assert(_stride > 0 && _stride <= 4);
```

Reply:
`assert(_stride >= 2 && _stride <= 4);` might be better?

Reply:
what about the number `3` 😶

Reply:
`assert(_stride == 2 || _stride == 4);`

Reply:
> `assert(_stride == 2 || _stride == 4);`

ok